### PR TITLE
Tickets/sitcom 1234

### PR DESCRIPTION
**Pull request by boutigny**
I think that I have implemented all the changes requested in SITCOM-1234

### CODE REVIEW

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py:1]: Hi, @boutigny - could you rename this file to something like `sitcomtn095.py` to match the other two files in the same folder?

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+    reference_time: str
+        slew stop time in UTC
+    df_ims: pandas data frame
+        IMS (Independent Measurement System of the M1M3) data
+    imsColumn: str
+        specific IMS column to analyze in this function call
+    correctedVariable:
+        column value corrected for the value at the end of the window, determined before the function call
+    req: float
+        tolerance for IMS value to be within requirement. We will apply it to the RMS and bias of the value, with respect to the value at reference_t + post_delta_t
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: We usually want docstring lines to be shorter than 80 characters. May I ask you to fix that? I was hoping Black would handle it.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+    meanAtReq: float
+        value of the bias (in absolute value) of the column value, using a rolling check, after the slew stop
+    settleTime: float
+        computed as latest time in which the IMS value has gone above the requirement in RMS or mean
+    """
+    ## recomputing RMS for the whole range since T0
+    rolling = 20
+    time_array = df_ims.index
+    slew_stop = Time(referenceTime).unix
+    iT0 = np.argmin(
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Do we know what `iT0` means? If so, could we have a more intuitive name for it to improve code readability?

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    meanAtReq = mean[iT1 - iT0]
+    krms = [index for index, x in enumerate(rms) if np.abs(x) >= req]
+    kmean = [index for index, x in enumerate(mean) if np.abs(x) >= req]
+    if (all(x < req for x in rms[iT1 - iT0 : -1])) and (
+        all(x < req for x in mean[iT1 - iT0 : -1])
+    ):
+        # all values of RMS and mean are OK since 1 s after slew stop
+        if verbose:
+            log_message(f"{imsColumn} Test PASSED", f)
+        if (krms == []) and (
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: I would add an empty line between `if` statements. This gives the code some "air" and lets the eyes know when the code is "slightly changing the subject." Kind of a weak visual clue.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+                - time_array[iT0]
+            )
+        settleTime = settleTime.total_seconds()
+        PF = False
+    log_message(f"settleTime:{settleTime}", f)
+    return PF, rmsAtReq, meanAtReq, settleTime
+
+
+def compute_settle_time(
+    df_ims,  # input data frame
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Let's remove these comments from the argument definition. Any comment here should be, instead, a documentation inside the docstring.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    referenceTime="2023-06-01T06:00:0Z",  # time for slew stop (T0)
+    lo_delta_t=5,  # in seconds
+    hi_delta_t=30,  # in seconds
+    imsColumn="xPosition",  # IMS column
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: We have a mix of `snake_case` and `semiCamelCase` variables. Could you have them all being `snake_case`?

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+    rmsReq=2e-3,  # requirement in appropriate units
+    chi2prob=0.999,  # confidence level for IMS variable wrt to long term value and variance to agree
+    f="SITCOM_1172.log",
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Can we rename this argument/variable to something more self-explanatory? Maybe `log_file`?

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+    reference_time: str
+        slew stop time in UTC
+    lo_delta_t: int
+        time window in seconds BEFORE the slew stop to retrieve data
+    hi_delta_t: int
+        time window in seconds AFTER the slew stop to retrieve data
+    imsColumn: str
+        specific IMS column to analyze in this function call
+    rmsReq: float
+        tolerance for IMS value to be within requirement. We will apply it to the RMS and bias of the value, with respect to the value at reference_t + post_delta_t
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Same thing here. We would like to have docstrings limited to 80 chars length max.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+    elif "Rotation" in imsColumn:
+        units = "deg"
+        ylimMax = rmsReq + 0.0001
+    else:
+        print("Unidentified column")
+        return -1
+
+    settleTime = False
+
+    T0 = pd.to_datetime(referenceTime)  # this is slew stop
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: I suggest replacing `T0` with `slew_stop_t`. Then the variable name becomes more self-explanatory.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+    delta_window = [
+        pd.Timedelta(lo_delta_t, "seconds"),
+        pd.Timedelta(hi_delta_t, "seconds"),
+    ]
+    # zoom around the T0 of interest
+    TZoom = [T0 - delta_window[0], T0 + delta_window[1]]
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Following the suggestion above, this one would be `zoom_t`. Or `t_zoom` if you prefer to have the `t` as a prefix. As long as the time variables have some sort of consistency to help readers quickly identify them as timestamps.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+
+    # targetVariablePlot takes the data frame for the complete plot range
+    targetVariablePlot = df_ims[imsColumn][TZoom[0] : TZoom[1]]
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: We should remove the spaces between `:`. See [this sections of PEP8 for more examples](https://peps.python.org/pep-0008/#whitespace-in-expressions-and-statements).

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+                max(time_array[iT0 + krms[-1]], time_array[iT0 + kmean[-1]])
+                - time_array[iT0]
+            )
+        settleTime = settleTime.total_seconds()
+        PF = False
+    log_message(f"settleTime:{settleTime}", f)
+    return PF, rmsAtReq, meanAtReq, settleTime
+
+
+def compute_settle_time(
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: This function is much easier to read than the function above, thanks to the empty lines it has in the middle of the code. If would be great if we could have the same blank lines in the rest of the code for both consistency and to improve readability.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+    # check the chi2 at each step using rollingCheck as the number of consecutive instances in which
+    # chi2 has to be under the critical value
+    # or rms and bias be both already 10% of requirement
+    PFCheck = (chi2 < crit) | ((rms < 0.1 * rmsReq) & (mean < 0.1 * rmsReq))
+    # PFCheck = (rms < 0.2 * rmsReq) & (mean < 0.5 * rmsReq)
+    rollingCheck = 30
+    stabilityCheck = (
+        PFCheck.rolling(rollingCheck).apply(lambda s: s.all()) > 0
+    )  # true if rollingCheck consecutive true values of PFcheck
+    if len(stabilityCheck[stabilityCheck == True]) <= rollingCheck:  ## == 0:
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: I would add a space before this `if`.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+    else:
+        n = 1
+        while (
+            rms[stabilityCheck[stabilityCheck == True].index[n + rollingCheck]] > rmsReq
+        ):
+            settleTime = stabilityCheck[stabilityCheck == True].index[n + rollingCheck]
+            n = n + 1
+        # if settleTime < referenceTime:
+        #    settleTime = referenceTime
+    settleInterval = -1
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: And another space before this line.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+def log_message(mess, f):
+    print(mess)
+    try:
+        print(mess, file=f)
+    except:
+        print("Could not write to log file")
+    return
+
+
+def get_slew_times(events, i_slew):
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: This function needs a docstring.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 15]
```diff
@@ -0,0 +1,657 @@
...
+    t0 = Time(events[i_slew].begin, format="isot", scale="utc")
+    t0 = pd.to_datetime(t0.value, utc=True)  # astropy Time to Timestamp conversion
+    t1 = Time(events[i_slew].end, format="isot", scale="utc")
+    t1 = pd.to_datetime(t1.value, utc=True)  # astropy Time to Timestamp conversion
+
+    return (t0, t1)
+
+
+def get_IMS_data(events, i_slew, postPadding):
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: This one needs a docstring too.

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py [hunk 16]
```diff
@@ -0,0 +1,657 @@
...
+    settleTimeXPosAgg = []
+    settleTimeYPosAgg = []
+    settleTimeZPosAgg = []
+    settleTimeXRotAgg = []
+    settleTimeYRotAgg = []
+    settleTimeZRotAgg = []
+
+    ignoreList = [92, 120, 274]  # these are specific seqNums to ignore
+
+    for i in range(len(blockEvents)):
```
Comment (b1quint) [inline python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: This for loop could benefit from some empty lines within it to help "group" lines that are closely related.

**Review by b1quint — CHANGES_REQUESTED** [review]
I added a few comments, see above. Hopefully this  will now make the comments show up?

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py:1]: Done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 1 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 3 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 2 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: renamed to something more explicit

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 12 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 13 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: Also removed the 2 commented lines

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 14 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 15 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 16 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 4 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: removed

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 5 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: I think that I have now converted all (most) of them

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 6 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: renamed

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 8 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 9 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: done

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 11 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: ok, I tried to add some extra blank lines to improve code readibility

File: python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py (hunk 7 above)
Comment (boutigny) [reply python/lsst/sitcom/vandv/m1m3/SITCOMTN-095_settling_time_block_analysis.py]: I think it is ok everywhere now

**Review by b1quint — APPROVED** [review]
This looks better now. Thank you very much!

**Pull request merged**